frontend/apps/reader/modules/readertoc.lua

```diff
@@ -436,21 +446,32 @@ function ReaderToc:isChapterEnd(cur_pageno)
 end
 
 function ReaderToc:getChapterPagesLeft(pageno)
-    --if self:isChapterEnd(pageno) then return 0 end
-    local next_chapter = self:getNextChapter(pageno)
-    if next_chapter then
-        next_chapter = next_chapter - pageno - 1
+    local flow = self.ui.document:getPageFlow(pageno)
+    -- Count pages until new chapter
+    local pages_left = 0
+    local test_page = self.ui.document:getNextPage(pageno)
+    while test_page > 0 do
+        pages_left = pages_left + 1
+        if self:isChapterStart(test_page) then
+            return pages_left
+        end
+        test_page = self.ui.document:getNextPage(test_page)
     end
-    return next_chapter
 end
 
 function ReaderToc:getChapterPagesDone(pageno)
     if self:isChapterStart(pageno) then return 0 end
-    local previous_chapter = self:getPreviousChapter(pageno)
-    if previous_chapter then
-        previous_chapter = pageno - previous_chapter
+    local flow = self.ui.document:getPageFlow(pageno)
+    -- Count pages until chapter start
+    local pages_done = 0
+    local test_page = self.ui.document:getPrevPage(pageno)
+    while test_page > 0 do
+        pages_done = pages_done + 1
+        if self:isChapterStart(test_page) then
+            return pages_done
+        end
+        test_page = self.ui.document:getPrevPage(test_page)
     end
-    return previous_chapter
 end
```

Comment:
Is there something wrong with the existing approach if `!hasFlows`? If not, can we keep it in this case? ;).

Reply:
Nothing wrong, I just like it better to have a single solution than lots of `if`s.

Reply:
`getPreviousChapter` & co happen to be faster by using the ticks directly, though, instead of walking page-by-page ;)

Reply:
>     * but I and other users should not see the complicated Goto description and all the new stuff when hide_non_linear is not true : everything should look as before when a user has not opt-in for the non-linear new behaviour

That's the case. Everything (but the new menu item) should look the same as it does now when disabled.

>     * (issue with bookmarks which could have some the normal page, some the tweaked non-linear page number, depending on whether they were made in one mode or the other - but that's a more general issue with our messy Bookmark stuff)

Right, but that also happens now if I make some bookmarks with large font and others with small font.

>     * I think the new menu should go into the first top menu (with TOC, Bookmarks, Reference pages) : and should not be shown when the document does not have any non-linear stuff. It should appear there only when there are (like Reference pages do)

Hmm...  But I want to be able to enable it for all books by default, even if the current book does not have it. Also, I thought that this is something that affects the formatting of the book and the reading experience, and it's not a tool for navigation, so it seems to fit better the second menu, doesn't it?